App crashed during some download failure

**App Version:** 1.57b

**Problem:** The app crashed in certain download failures (either by pressing download button or just getting a chapter from fanfiction). 

**Root cause:** when the network is unstable, and the underlying `Jsoup` in some cases will convert checked `IOException` to `UncheckedIOException`, which then bypasses all error handling in the app and causes the app crashed.

Example stacktrace:
```java
E/AndroidRuntime: FATAL EXCEPTION: Story Loader Thread
    Process: com.spicymango.fanfictionreader, PID: 20636
    java.lang.RuntimeException: An error occurred while executing doInBackground()
        at android.support.v4.content.ModernAsyncTask$3.done(ModernAsyncTask.java:161)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
...
     Caused by: org.jsoup.UncheckedIOException: java.io.IOException: Underlying input stream returned zero bytes
        at org.jsoup.parser.CharacterReader.bufferUp(CharacterReader.java:60)
        at org.jsoup.parser.CharacterReader.<init>(CharacterReader.java:35)
        at org.jsoup.parser.CharacterReader.<init>(CharacterReader.java:39)
        at org.jsoup.parser.TreeBuilder.initialiseParse(TreeBuilder.java:35)
        at org.jsoup.parser.HtmlTreeBuilder.initialiseParse(HtmlTreeBuilder.java:66)
        at org.jsoup.parser.TreeBuilder.parse(TreeBuilder.java:44)
        at org.jsoup.parser.Parser.parseInput(Parser.java:39)
        at org.jsoup.helper.DataUtil.parseInputStream(DataUtil.java:151)
        at org.jsoup.helper.HttpConnection$Response.parse(HttpConnection.java:832)
        at org.jsoup.helper.HttpConnection.get(HttpConnection.java:289)
        at com.spicymango.fanfictionreader.activity.reader.StoryDisplayActivity$FanFictionLoader.getStoryFromSite(StoryDisplayActivity.java:638)
        at com.spicymango.fanfictionreader.activity.reader.StoryLoader.loadInBackground(StoryLoader.java:96)
        at com.spicymango.fanfictionreader.activity.reader.StoryLoader.loadInBackground(StoryLoader.java:15)
        at android.support.v4.content.AsyncTaskLoader.onLoadInBackground(AsyncTaskLoader.java:302)
        at android.support.v4.content.AsyncTaskLoader$LoadTask.doInBackground(AsyncTaskLoader.java:57)
        at android.support.v4.content.AsyncTaskLoader$LoadTask.doInBackground(AsyncTaskLoader.java:45)
        at android.support.v4.content.ModernAsyncTask$2.call(ModernAsyncTask.java:138)
        at java.util.concurrent.FutureTask.run(FutureTask.java:237)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607) 
        at java.lang.Thread.run(Thread.java:761) 
     Caused by: java.io.IOException: Underlying input stream returned zero bytes
...
```

--- a/fanfictionReader/src/main/java/com/spicymango/fanfictionreader/activity/reader/StoryDisplayActivity.java
+++ b/fanfictionReader/src/main/java/com/spicymango/fanfictionreader/activity/reader/StoryDisplayActivity.java
@@ -56,6 +56,7 @@ import com.spicymango.fanfictionreader.provider.StoryProvider;
 import com.spicymango.fanfictionreader.services.LibraryDownloader;
 import com.spicymango.fanfictionreader.util.AsyncPost;
 import com.spicymango.fanfictionreader.util.FileHandler;
+import com.spicymango.fanfictionreader.util.JsoupUtil;
 import com.spicymango.fanfictionreader.util.Sites;
 import com.spicymango.fanfictionreader.util.adapters.TextAdapter;
 
@@ -634,8 +635,8 @@ public class StoryDisplayActivity extends AppCompatActivity implements LoaderCal
 			builder.appendEncodedPath(Long.toString(storyId));
 			builder.appendEncodedPath(Integer.toString(currentPage));
 			builder.appendEncodedPath("");
-			
-			Document document = Jsoup.connect(builder.toString()).timeout(10000).get();
+
+			Document document = JsoupUtil.safeGet(builder.toString());
 			
 			if (data.getTotalChapters() == 0) {
 				Element title = document.select("div#content div b").first();

--- a/fanfictionReader/src/main/java/com/spicymango/fanfictionreader/services/DownloaderFactory.java
+++ b/fanfictionReader/src/main/java/com/spicymango/fanfictionreader/services/DownloaderFactory.java
@@ -12,6 +12,7 @@ import com.spicymango.fanfictionreader.activity.Site;
 import com.spicymango.fanfictionreader.provider.SqlConstants;
 import com.spicymango.fanfictionreader.provider.StoryProvider;
 import com.spicymango.fanfictionreader.util.FileHandler;
+import com.spicymango.fanfictionreader.util.JsoupUtil;
 import com.spicymango.fanfictionreader.util.Parser;
 import com.spicymango.fanfictionreader.util.Sites;
 import com.spicymango.fanfictionreader.util.Story;
@@ -268,7 +269,7 @@ class DownloaderFactory {
 		public void downloadChapter() throws IOException, ParseException, StoryNotFoundException {
 			final String url = "https://www.fanfiction.net/s/" + mStoryId + "/"
 					+ mCurrentPage + "/";
-			final Document document = Jsoup.connect(url).timeout(10000).userAgent("Mozilla/5.0").get();
+			final Document document = JsoupUtil.safeGet(url, "Mozilla/5.0");
 
 			// On the first run, update the mStory variable
 			if (mCurrentPage == 1) {

--- a/fanfictionReader/src/main/java/com/spicymango/fanfictionreader/util/JsoupUtil.java
+++ b/fanfictionReader/src/main/java/com/spicymango/fanfictionreader/util/JsoupUtil.java
@@ -1,0 +1,41 @@
+package com.spicymango.fanfictionreader.util;
+
+import android.support.annotation.NonNull;
+
+import org.jsoup.Connection;
+import org.jsoup.Jsoup;
+import org.jsoup.UncheckedIOException;
+import org.jsoup.nodes.Document;
+
+import java.io.IOException;
+
+public class JsoupUtil {
+	/**
+	 * A helper to use Jsoup get().
+	 *
+	 * It converts any underlying unchecked {@link UncheckedIOException} to checked {@link IOException}
+	 * and throws the checked on instead.
+	 *
+	 * Use case: application codes have been written to handle and recover from checked IOException.
+	 * This helper ensures the recovery will be equally applied to unchecked ones. Otherwise,
+	 * the application will crash.
+	 *
+	 */
+	public static Document safeGet(@NonNull String uri, String userAgent) throws IOException {
+		try {
+			Connection conn = Jsoup.connect(uri).timeout(10000);
+			if (userAgent != null) {
+				conn.userAgent(userAgent);
+			}
+
+			return conn.get();
+		} catch (UncheckedIOException uioe) {
+			throw uioe.ioException();
+		}
+	}
+
+    public static Document safeGet(@NonNull String uri) throws IOException {
+	    return safeGet(uri, null);
+    }
+
+}

